ERCS/erc-7700.md

```diff
@@ -0,0 +1,549 @@
+
+```js
+/* CAIP-10 identifier */
+const caip10 = `eip155:${chainId}:${wallet}`
+```
+
+##### `password`
+`password` is an optional private field and it must be prompted from the user by the client; this field allows users to change data signers for a given `username`.
+```js
+/* Secret derived key identifier */ 
+// Clients must prompt the user for this
+const password = 'key1'
+```
+
+##### `sigKeygen`
+The data signer must be derived from the owner or manager keys of a node. Message payload for the required `sigKeygen` must then be formatted as:
+
+```text
+Requesting Signature To Generate Keypair(s)\n\nOrigin: ${username}\nProtocol: ${protocol}\nExtradata: ${extradata}
+```
+
+where the `extradata` is calculated as follows,
+
+```solidity
+// Calculating extradata in keygen signatures
+bytes32 extradata = keccak256(
+    abi.encodePacked(
+        pbkdf2(
+            password, 
+            salt, 
+            iterations
+        ), // Stretch password with PBKDF2
+        wallet
+    )
+)
+```
+
+where `PBKDF2` - with `keccak256(abi.encodePacked(username))` as salt and the `iterations` count is fixed to `500,000` for brute-force vulnerability protection.
+
+```js
+/* Definitions of salt and iterations in PBKDF2 */
+let salt = keccak256(abi.encodePacked(username));
+let iterations = 500000; // 500,000 iterations
+```
+
+The remaining `protocol` field is a protocol-specific identifier limiting the scope to a specific protocol represented by a unique contract address. This identifier cannot be global and must be uniquely defined for each implementating `contract` such that:
+
+```js
+/* Protocol identifier in CAIP-10 format */
+const protocol = `eth:${chainId}:${contract}`;
+```
+
+With this deterministic format for signature message payload, the client must prompt the user for the ethereum signature. Once the user signs the messages, the `keygen()` function can derive the data signer keypair. 
+
+#### 2. Sign Data
+Since the derived signer is wallet-specific, it can 
+
+- sign batch data for multiple keys for a given node, and 
+- sign batches of data for multiple nodes owned by a wallet
+
+simultaneously in the background without ever prompting the user. Signature(s) `dataSig` accompanying the off-chain calldata must implement the following format in their message payloads:  
+
+```text
+Requesting Signature To Update Off-Chain Data\n\nOrigin: ${username}\nData Type: ${dataType}\nData Value: ${dataValue}
+```
+
+where `dataType` parameters are protocol-specific and formatted as object keys delimited by `/`. For instance, if the off-chain data is nested in keys as `a > b > c > field > key`, then the equivalent `dataType` is `a/b/c/field/key`. For example, in order to update off-chain ENS record `text > avatar` and `address > 60`, `dataType` must be formatted as `text/avatar` and `address/60` respectively.
+ 
+#### 3. Approve Data Signer
+To further save on gas costs, we do not store the `dataSigner` on L1. Instead, the clients must
```

Comment:
> To further save on gas costs, we do not store the `dataSigner` on L1.

This content belongs in the Rationale section. Keep the Specification section as focused as possible to make it easier for implementers to follow.

Reply:
Fixed in https://github.com/ethereum/ERCs/pull/404/commits/363fc77d6fbba3919a18aaab2ba757e9af9fae42